[GEOT-6740] Shapefile reader fails on lines and polygons having Z and no optional M

taba90 opened the pull request:
Jira ticket: https://osgeo-org.atlassian.net/browse/GEOT-6740?searchSessionId=b0bbfbee-f28d-4329-957d-fd1a24556038&searchObjectId=53422&searchContainerId=10001&searchContentType=issue

- this pr fixes and issue with polygon/lines shafile of type ARCZ with no optional M
- no test have been added, instead the two one already existing for this use case have been modified to use data that really doesn't have the optional M.


## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [x] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer (there is an automatic PR check verifying this, check this when it turns green).

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):
- [x] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message(s) must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [x] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

aaime commented on modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/shp/MultiLineHandler.java:
```diff
@@ -221,10 +221,13 @@ public Object read(ByteBuffer buffer, ShapeType type, boolean flatGeometry) {
             }
         }
-        if ((shapeType == ShapeType.ARCZ || shapeType == ShapeType.ARCM) && !flatGeometry) {
+        if ((doubleBuffer.hasRemaining() && (shapeType == ShapeType.ARCZ)
```
Points and multipoints have not been covered by the changes, looking at their code, is seems they would be affected by the same issue. Or not?

tlbeale replied:
Multipoints also have the same problem. Points have already been fixed up in master by GEOT-6729 (commit 821d9add2f8de)

tlbeale commented:
Hi,

It turns out I have been working on the same problem as you. I was just going to submit a PR under GEOT-6683 when I saw your one. I have added test cases that highlight the problem, if that's any help to you. You can see/cherry-pick my changes here (the unsquashed version is spread over 4 commits).
https://github.com/tlbeale/geotools/commits/GEOT-6683-unsquashed

Alternatively, I can turn my changes into a PR if that's easier. But I thought it was confusing having 2 PRs for the same thing...

tlbeale commented on modules/plugin/shapefile/src/test/java/org/geotools/data/shapefile/shp/ZMHandlersTest.java:
```diff
@@ -155,15 +155,15 @@ public void testReadMLine() throws ShapefileException, IOException {
     @Test
     public void testReadZLine() throws ShapefileException, IOException {
-        URL url = TestData.url(ShapefileDataStore.class, "mzvalues/zlines.shp");
+        URL url = TestData.url(ShapefileDataStore.class, "mzvalues/linez.shp");
```
I think it needed to be the last shape in the file being read to trigger the exception (which is the difference between overrunning the buffer vs just reading random data from the next shape in the file). The existing tests were reading the Z-shapes without M-values and passing because they only ever read the _first_ shape. @taba90 changed the file to only contain a single shape, which would then trigger the exception. (For GEOT-6683 I took a slightly different approach and used the existing file, and added a new test that read _all_ the shapes in the file)

taba90 commented:
@aaime thanks for the review, I've added the MultiPoints handling and took back the shapefile I've removed initially.
@tlbeale thanks for your comment I've picked your commits regarding the test class and squashed into one.
As a result we have now test for the three variants: ArcTypeM, ArcTypeZ with optional M and ArcTypeZ with no optional M.
The existing data meant to test the third use case were actually having the M dimension. I've thus added new ones.

ianturton commented:
The existing datasets were supposed to test all the legal variants - I'm
not sure it's legal to have M with out Z (or possibly the other way round)
- do we know where these Just Z and Just M shapefiles are coming from?

Ian

On Tue, 24 Nov 2020 at 09:40, taba90 <notifications@github.com> wrote:

> @aaime <https://github.com/aaime> thanks for the review, I've added the
> MultiPoints handling and took back the shapefile I've removed initially.
> @tlbeale <https://github.com/tlbeale> thanks for your comment I've picked
> your commits regarding the test class and squashed into one.
> As a result we have now test for the three variants: ArcTypeM, ArcTypeZ
> with optional M and ArcTypeZ with no optional M.
> The existing data meant to test the third use case were actually having
> the M dimension. I've thus added new ones.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/geotools/geotools/pull/3219#issuecomment-732777946>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA55IKJ43FGBOQO4WNOG53SRN5QLANCNFSM4TYNRCWA>
> .
>


-- 
Ian Turton

aaime commented:
@ianturton I skimmed throught the spec the other day, my current understanding is that:
* One can have just M, if the type is a <geomType>M
* One can have just Z, or both Z and M, if they type is a <geomType>Z

ianturton commented:
Possibly, it's been a while since I read the spec. - So do we know what the
difference between these new files and the old ones that I thought covered
those types (that I made in QGis)?

Ian

On Tue, 24 Nov 2020 at 16:17, Andrea Aime <notifications@github.com> wrote:

> @ianturton <https://github.com/ianturton> I skimmed throught the spec the
> other day, my current understanding is that:
>
>    - One can have just M, if the type is a M
>    - One can have just Z, or both Z and M, if they type is a Z
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geotools/geotools/pull/3219#issuecomment-733083471>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA55IOUECGE6PSUOLZLZ33SRPMCBANCNFSM4TYNRCWA>
> .
>


-- 
Ian Turton

taba90 commented:
@ianturton thanks for your comment. While running test with the old files it turned out that they actually have additional bytes for the M. Also looks to me that when shapefile of type ARCZ  are created with QGIS it adds to them also the M value: if one creates a shapefile from the create new shapefile button indeed the allowed additional dimensions between wich one can choose when a shapefile is created are none, Z + M and M
![CatturaQGIS](https://user-images.githubusercontent.com/46672505/100198784-974d6480-2efc-11eb-8e27-01b55d4627c1.PNG). The newly added test file have been create trough gdal.

aaime commented:
Looks good to me, thanks @taba90 and @tlbeale
